Sources/SpeziHealthKitBulkExport/SpeziHealthKitBulkExport.docc/BulkHealthExporter.md

```diff
@@ -0,0 +1,97 @@
+
+SPDX-FileCopyrightText: 2025 Stanford University and the project authors (see CONTRIBUTORS.md)
+
+SPDX-License-Identifier: MIT
+-->
+
+Export large amounts of historical Health data
+
+## Overview
+
+The ``BulkHealthExporter`` enables and coordinates large-scale support of historical HealthKit data.
+
+The Bulk Export API is built around the concept of Export Sessions (``BulkExportSession``), which implement and handle the Health data export processing. 
+Sessions keep track of their pending and already-completed work, including across multiple app launches, ensuring that even for sample types with a very high number of samples a previously-started export can continue without issues if the app is terminated during the export.
+
+Export Sessions are created using ``BulkHealthExporter/session(_:for:startDate:endDate:batchSize:using:startAutomatically:)``, and consist of the following components:
+- A stable identifier, which is used to keep track of the session, persist its state, and restore it across app launches.
+- A set of to-be-exported sample types.
+- A ``BatchProcessor``, which allows the app to process the individual batches of fetched samples.
+
+This structure allows the Bulk Export API to be used in a variety of ways, for different kinds of export operations. (See below for an example.)
+
+In order to perform a Health data export, an app simply calls the ``BulkHealthExporter/session(_:for:startDate:endDate:batchSize:using:startAutomatically:)`` function once at some point after the app's launch; this will either:
+- create and start a new session, if no matching session (based on the identifier) exists, or
+- restore and continue an existing session (e.g., from a previous launch of the app).
+
+It is safe to call this function multiple times and with the same input, even if a previously-created upload has already been completed.
+The session will internally keep track of its creation date, and will only ever export samples up to that date.
+This allows an app to e.g. use the `CollectSample` API to continuously fetch and collect new Health samples, and use the ``BulkHealthExporter`` to do a one-time export operation of historical Health data.
+
+The ``BulkHealthExporter/session(_:for:startDate:endDate:batchSize:using:startAutomatically:)`` function will, when a ``BulkExportSession`` is first created, also return an `AsyncStream` which can be used to access the individual results of the session's ``BatchProcessor``.
+
+- Important: Ensure that your app has sufficient HealthKit access permissions before starting bulk export sessions. The session itself will *not* prompt the user for access; instead, it will fail to fetch and process any sample types for which no HealthKit permission is granted.
+
+It is possible to ``BulkExportSession/pause()`` an export session, which can then be resumed using the ``BulkExportSession/start(retryFailedBatches:)`` function.
+
+
+### Example: Bulk-Upload of Historical Health Data to Firebase
```

Comment:
It would be great to also add an example of a `BatchProcessor` that uses an output type and how to properly use the async sequences.

Reply:
added a second one that is essentially the MHC BatchProcessor minus the compression. also neatly demonstrates that you can combine this API with eg HealthKitOnFHIR to process the samples into another format

Reply:
Thank you!